### ISSUE TITLE
Load plugins in parallel?

### ISSUE THREAD

**Issue by brettwillis** [fastify/help#897]
In some cases I have plugins that are not dependent on eachother and take many seconds to load. It seems fastify loads them in sequence, waiting for one to finish before starting the next.

When searching the docs it's hard to tell that this behaviour was the case. The only mention was "avvio maintains load and close order".

In the case mentioned above it's a big waste of time loading because they can be loaded in parallel as they are not dependent.

In fact, in my experience, plugins can almost always load in parallel, and if not I specify plugin dependencies.

Is there a way to specify a plugin graph which loads in parallel?

**Comment by Eomm** [comment]
> Is there a way to specify a plugin graph which loads in parallel?

No, there is no ATM

If the plugin is slow to start, I would load the resource in a lazy loading way to keep the start time as fast as possible.

It would be a cool idea to think if we can ship such a feature on `fastify-plugin`

**Comment by metcoder95** [comment]
>It would be a cool idea to think if we can ship such a feature on fastify-plugin

How it would look like? 🤔

**Comment by brettwillis** [comment]
What if there was a flag in the consctructor, like below, where `'automatic'` causes plugins to always load in parallel, provided they load after any specified "dependencies"? I'm referring to the `dependencies` option from `fastify-plugin`.

```ts
/**
 * @default 'sequential'
 */
pluginLoadOrder?: 'sequential' | 'automatic';
```

However I wonder how it would work if you need to make a third-party plugin depend on another third-party plugin (i.e. we aren't able to specify dependencies of a third-party plugin). You'd also need to also add a `dependencies` field to the `register()` options, which would add to any already specified dependencies.

**Comment by metcoder95** [comment]
I think `fastify-plugin` can provide some abstraction to that, in the sense of marking the dependencies required by the `plugin` being registered so a dependency graph can be altered or just a new leaf added to the graph.

Even though, if looking for truly parallel it requires some changes to `avvio` instead. Although `lazy loading` is something that can be abstracted within `fastify-plugin` at some extent but requires mostly user work at the plugin level itself 🤔

**Comment by brettwillis** [comment]
Regarding lazy loading, I assume you mean the plugin function's promise/done() called quickly, before the long running task completes?

For a scenario I have in mind, lazy loading (abstracted by `fastify-plugin` or not) is not really an option. 

<details><summary><i>why?</i></summary>
<p>
The long running task could be downloading/preparing data that is required to serve requests. And it's a serverless runtime environment where, once the app's container is "ready", it will start receiving requests. The result would be that initial requests to the container would see a long latency. In short, the app must wait for the plugin(s) to finish loading before becoming "ready".
</p>
</details> 

> if looking for truly parallel it requires some changes to avvio instead

Yeah I assumed that this would be the case. `avvio` would need to do all the plugin dependency graphs and wait for dependencies to finish before starting dependents.

**Comment by Uzlopak** [comment]
I assume, you would create a new Symbol, which fastify-plugin would set on the function, and then avvio would pick this up. 

I think an implementation should be already possible for simple cases without dependencies, by defining the pluginFn in callback style and call the done-callback early so that the avvio queue continues to process.

**Comment by brettwillis** [comment]
> why is your plugin slow? Does it connect to something, and then it adds a decorator? Does it add a route instead?

It's doing all sorts of things like downloading a dataset, connecting to a Pub/Sub subscription etc. Conceivably it can add all of its routes and decorators before waiting for those async tasks to finish. But they must finish before the app is ready.


However I'm kind of thinking a bit more generally. In general most plugins are only dependent on a small set of other plugins, not every other plugin registered beforehand. Even if plugins load quickly, within a second, if you have tens or hundreds of plugins the app will still load slowly, when unnecessarily doing sequential loading.

**Comment by metcoder95** [comment]
>Regarding lazy loading, I assume you mean the plugin function's promise/done() called quickly, before the long running task completes?

Yes

>I assume, you would create a new Symbol, which fastify-plugin would set on the function, and then avvio would pick this up.
>
>I think an implementation should be already possible for simple cases without dependencies, by defining the pluginFn in callback style and call the done-callback early so that the avvio queue continues to process.

Yeah, it was more or less what I had in mind. Mark them so `avvio` does not wait for them and does the rest of the job. We can stop all the queue processing as soon as one fails (basically some sort of `Promise.all` but with better control).

> - the plugin can't add decorators, hooks, etc...

I think they can still decorate the Fastify app and set hooks, just before the server starts listening. Anyways, we need all of the plugins to be loaded just before the app starts listening.

>However I'm kind of thinking a bit more generally. In general most plugins are only dependent on a small set of other plugins, not every other plugin registered beforehand. Even if plugins load quickly, within a second, if you have tens or hundreds of plugins the app will still load slowly, when unnecessarily doing sequential loading.

Generally speaking, for smaller plugins that do not require dependent on others, the proposal done by @Uzlopak sounds feasible, and maybe something we can experiment with. We can control it through a flag (e.g. `asyncStart`), which underneath it adds a `Symbol` that only `avvio` understands. For plugins with a more complex graph, we can think on something down the road

**Comment by Uzlopak** [comment]
Somehow i think we would need a new state when loading the plugin, and maybe allow a second callback. 

Currently if the pluginFn has three arguments, it is a callback plugin. So the plugin is finished loading, when the callback is called.

If the fourth argument is also a callback, then we could say, that the first callback is meaning that the plugin is loaded and the second callback means that the plugin is ready or started. So we could instruct avvio, that if the plugin is loaded, than continue loading the next plugin. And if the second callback is called, avvio marks it as finished and resolving the queue. 

If the plugin has only one callback, it would immediatly set loaded and finished. So no behaviour change. 

Lazy Loadig would not need a special symbol or so.

Also because the second callback delays the ready event of avvio, it would not result in starting fastify before the second done is called, so you could still add routes and hooks

Only drawback would be that you could get race conditions, e.g. when plugin A creates a route and plugin B is setting the same route. Then sometimes plugin A would throw an error and sometimes plugin B would throw and error. Potentially making debugging harder.

It is for me hard to explain.

**Comment by brettwillis** [comment]
> If the fourth argument is also a callback, then we could say, that the first callback is meaning that the plugin is loaded and the second callback means that the plugin is ready or started. So we could instruct avvio, that if the plugin is loaded, than continue loading the next plugin. And if the second callback is called, avvio marks it as finished and resolving the queue.

Sounds like a good enough workaround. Plugins would still be loaded sequentially, but any long-running async tasks would be allowed to finish while the loading chain continues in parallel? Not fully parallel and automatic (based declared dependencies) like `pluginLoadOrder: 'automatic'`, for example it depends on plugin implementation so if a third-party plugin doesn't defer it's async tasks then you can't make it load in parallel based on application-specific dependency chains. But it is definitely a huge improvement.

When a plugin has a dependency on another plugin, it is because it uses the decorators registered by that dependency. So the one callback should be called only after it's registered all it's decorators. The other to be called once any (optional) async tasks finish.

And just a suggestion, the *"I've registered my decorators but may have async work pending"* callback should probably be the second one, because it is optional. The *"I've finished everything and I'm ready"* callback should be the first one. Then it's bacwards compatible with existing plugins that only use the one callback.

So like this:

```js
const myNormalPlugin = function (instance, opts, done) {
  // register my decorators etc ...
  // calling done() implicitly calls next() if not already called
  done();
}
const myAsyncPlugin = function (instance, opts, done, next) {
  // register my decorators etc ...
  // plugin loading chain can continue
  next();

  // do some more async work...
  // I'm ready to start
  done();
}
```

Would the `asyncStart` flag be necessary then?

**Comment by mcollina** [comment]
This would be _very hard_ to achieve as it would need to happen in Avvio and it would touch quite complex code. Moreover, there are a lot of side effects to what has been described here.

If we were to implement something to fix this, it won't be a property of the loaded plugin but rather of the plugin currently being loaded. We should investigate how we could raise the number of (sub) plugins that are concurrently loaded (the `1` parameter in this [function](https://github.com/fastify/avvio/blob/07614c05ff599cef6f8a8c96ce0235d6ab9bc78a/plugin.js#L22)) . This is going to be hard with a lot of edge cases and I'm not exactly convinced this could be beneficial.

I think a deterministic bootstrap sequence is key to avoid race conditions and ultimately a good DX.

**Comment by brettwillis** [comment]
> I think a deterministic bootstrap sequence is key to avoid race conditions and ultimately a good DX.

What if the `next()` callback *requires* all decorators and routes to be registered, and Fastify enforced that none can be registered by the plugin after `next()` is called. Then there are no race conditions?

**Comment by mcollina** [comment]
I'm referring to how complex the Avvio codebase is. The last time we added a feature there, it required more than a month of full-time work from me to iron it out. I don't have the time to dedicate to it right now.

> With the two callbacks option (deferred ready state), what if the next() callback requires all decorators and routes to be registered, and Fastify enforced that none can be registered by the plugin after next() is called. Then there are no race conditions?

This is unfeasible due to how promises and callback support are interleaved.

**Comment by Uzlopak** [comment]
Imho This would need strong coupling between fastify and avvio, as you would need to touch fastify code and implement changes there.

The only reason for two callbacks is to delay the fastify ready event, until the plugin is finished loading. Other than that, you could already implement lazy loading, as described above by calling done early, but risking that fastify started before your plugins started completely. 

I think the fear of @mcollina is the side effect of implementing second callback would mean, that we maybe support such a feature, despite that we would only recommend this to devs who know what they are doing and know that they are getting into race condition world, if they add routes, hooks etc. to the fastify instance.

Unfortunately devs expect support for features. We would need to add a big fat disclaimer, saying that we dont support issues with lazy loading

**Comment by mcollina** [comment]
What I _thought_ a long time ago was to add an option to allow for a plugin-defined API to be called when `register(plugin, opts)` is called.

```js
const fp = require('fastify-plugi')
const myPlugin = fp(async function (app, opts) {
  const db = await opts.dbConnection
  app.decorate('db', db)
}, {
  onRegister (opts) {
    opts = { ...opts } // shallow clone

    // Note that this is a synchronous function
    opts.dbConnection = (async function connectToDB () {
      //...
    })()
    opts.dbConnection.catch(() => {}) // to avoid unhandledRejections

    return opts // we can completely override the options if we want
  }
})

fastify.register(app, opts)
```

I think this can be implemented without problems and it would solve the same need. The API would be deliberately not-nice as to signal it is territory for advanced users.

Note that this API could potentially cause memory leaks and this is why I ended up never implementing it.

**Comment by climba03003** [comment]
For the time being, you may try the logic similar to [`fastify-delay-request`](https://github.com/climba03003/fastify-delay-request)
All you need is a separate queue that should not block the server listen and requests.

```js
function (fastify, options, done) {
  // decorators, etc that requires sync operation
  
  // async process manage in separate queue
  fastify.asyncTask.add(async function() {
    // heavy task
  } /* maybe we can allow to configurate if the async task blocking or non-blocking */ )
  
  // do not block process and continue
  done()
}

// for implementing async task, all you need is a `onReady` hooks which is used as blocking server ready if needed
```

**Comment by metcoder95** [comment]
>Note that this API could potentially cause memory leaks and this is why I ended up never implementing it.

I think this kind of problem is the type of Pandora's box we'll open if we aim for advanced ways of loading plugins within `avvio` or `fastify` itself.

>```js
>  // async process manage in separate queue
>  fastify.asyncTask.add(async function() {
>    // heavy task
>  } /* maybe we can allow to configurate if the async task blocking or non-blocking */ )
>```

Maybe we can do the same but at the `register` level?

**Comment by brettwillis** [comment]
@climba03003 that's a great and simple idea, `fastify-delay-request` is almost exactly what is needed. The only difference is I'd prefer to start the tasks immediately, not wait until the app is ready. But that's just configuring the `autoStart` option on the `p-queue`.

> Maybe we can do the same but at the register level?

Yeah having this built-in to fastify would allow third-party plugins to all utilise the same system/technique. Although, they could just do the following inside the plugin itself. If that was just a plugin design guideline it could be enough?

```ts
const myPlugin = async (fastify) => {
  // Register my decordators etc

  const task: Promise<void> = // start some heay work...
  fastify.addHook('onReady', () => task);
};
```

Just thinking... would this cause problems with unhandled promise rejections, because promise callbacks are not registered until the onReady hook?

**Comment by climba03003** [comment]
> The only difference is I'd prefer to start the tasks immediately, not wait until the app is ready.

Yes, the idea is almost the same.

> Although, they could just do the following inside the plugin itself.

Not really, `onReady` hook run in sequential order. You need an extra queue for parallel staff.

> Just thinking... would this cause problems with unhandled promise rejections

I believe no, because you return a `Promise` and `fastify` will attach `.then` for you.
